core/src/main/java/org/axonframework/queryhandling/SimpleQueryBus.java

```diff
@@ -228,103 +227,48 @@ private boolean unsubscribe(String queryName,
             SubscriptionQueryMessage<Q, I, U> query,
             SubscriptionQueryBackpressure backpressure) {
-        MessageMonitor.MonitorCallback monitorCallback = messageMonitor.onMessageIngested(query);
-        SubscriptionQueryMessage<Q, I, U> interceptedQuery = intercept(query);
-        delayedUpdates.putIfAbsent(interceptedQuery, new CopyOnWriteArrayList<>());
-        MonoWrapper<QueryResponseMessage<I>> initialResult =
-                MonoWrapper.create(monoSink -> initialResult(interceptedQuery, monitorCallback, monoSink));
-        FluxWrapper<SubscriptionQueryUpdateMessage<U>> updates =
-                FluxWrapper.create(fluxSink -> updates(interceptedQuery, fluxSink), backpressure);
 
-        return new DefaultSubscriptionQueryResult<>(initialResult.getMono(), updates.getFlux());
+        MonoWrapper<QueryResponseMessage<I>> initialResult = MonoWrapper.create(monoSink -> query(query)
+                .thenAccept(monoSink::success)
+                .exceptionally(t -> {
+                    monoSink.error(t.getCause());
+                    return null;
+                }));
+
+        EmitterProcessor<SubscriptionQueryUpdateMessage<U>> processor = EmitterProcessor.create();
+        FluxSink<SubscriptionQueryUpdateMessage<U>> sink = processor.sink(backpressure.getOverflowStrategy());
+        sink.onDispose(() -> updateHandlers.remove(query));
+        updateHandlers.computeIfAbsent(query, k -> new CopyOnWriteArrayList<>())
+                      .add(new FluxSinkWrapper<>(sink));
+
+        return new DefaultSubscriptionQueryResult<>(initialResult.getMono(), processor.replay().autoConnect());
     }
 
     @SuppressWarnings("unchecked")
     @Override
     public <U> void emit(Predicate<SubscriptionQueryMessage<?, ?, U>> filter,
                          SubscriptionQueryUpdateMessage<U> update) {
-        scheduleAfterInitial(filter, (query, fluxSink) -> doEmit(query, fluxSink, update));
         updateHandlers.keySet()
                       .stream()
                       .filter(sqm -> filter.test((SubscriptionQueryMessage<?, ?, U>) sqm))
-                      .forEach(query -> {
-                          FluxSinkWrapper<?> updateHandler = updateHandlers.get(query);
-                          doEmit(query, updateHandler, update);
-                      });
+                      .forEach(query -> updateHandlers.get(query)
+                                                      .forEach(uh -> doEmit(query, uh, update)));
     }
 
     @Override
     public void complete(Predicate<SubscriptionQueryMessage<?, ?, ?>> filter) {
-        scheduleAfterInitial(filter, (query, fluxSink) -> fluxSink.complete());
         updateHandlers.keySet()
                       .stream()
                       .filter(filter)
-                      .forEach(query -> updateHandlers.remove(query).complete());
+                      .forEach(query -> updateHandlers.remove(query).forEach(FluxSinkWrapper::complete));
```

Comment:
I'm wondering what could happen if this code is executed concurrently (in the end, two events could trigger a completion of the updateHandler.
Instead of `.forEach(query -> updateHandler.remove(query))`, you could do something in the lines of: `.map(updateHandlers::remove).filter(nonNull).flatMap(List::stream).forEach(FluxSinkWrapper::complete)`

Reply:
You're right, it is even more dangerous if `updateHandler` is removed during the emitting.